crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,501 @@
+        input_wasm: &'a [u8],
+    ) -> crate::Result<()> {
+        self.write_code(ast, nodeidx, range, newfunc, operators, input_wasm)
+    }
+
+    /// Encoding discriminator for the AST nodes, it calls
+    /// the corresponding methods depending on the node type
+    ///
+    fn write<'a>(
+        &self,
+        ast: &AST,
+        nodeidx: usize,
+        newfunc: &mut Function,
+        operators: &Vec<OperatorAndByteOffset>,
+        input_wasm: &'a [u8],
+    ) -> crate::Result<()> {
+        let node = &ast.nodes[nodeidx];
+
+        match node {
+            Node::IfElse(then, alternative, ty) => {
+                self.write_if_else(
+                    ast,
+                    nodeidx,
+                    then,
+                    alternative,
+                    newfunc,
+                    operators,
+                    input_wasm,
+                    ty,
+                )?;
+            }
+            Node::Code { range } => {
+                self.write_code(ast, nodeidx, *range, newfunc, operators, input_wasm)?;
+            }
+            Node::Loop(body, ty) => {
+                self.write_loop(ast, nodeidx, body, newfunc, operators, input_wasm, ty)?
+            }
+            Node::Block(body, ty) => {
+                self.write_block(ast, nodeidx, body, newfunc, operators, input_wasm, ty)?
+            }
+            Node::Root(body) => {
+                for ch in body {
+                    self.write(ast, *ch, newfunc, operators, input_wasm)?;
+                }
+                // Closing end
+                newfunc.instruction(&Instruction::End);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ASTWriter for AST {
+    /* It has the default implementation */
+}
+
+impl AST {
+    /// Build a new empty AST
+    pub fn new(
+        root: usize,
+        nodes: Vec<Node>,
+        ifs: Vec<usize>,
+        loops: Vec<usize>,
+        blocks: Vec<usize>,
+    ) -> Self {
+        AST {
+            root,
+            nodes,
+            ifs,
+            loops,
+            blocks,
+        }
+    }
+
+    /// Returns true if the AST has if-else nodes
+    ///
+    pub fn has_if(&self) -> bool {
+        !self.ifs.is_empty()
+    }
+
+    /// Returns true if the AST has loop nodes
+    ///
+    pub fn has_loop(&self) -> bool {
+        !self.loops.is_empty()
+    }
+
+    /// Returns true if the AST has block nodes
+    ///
+    pub fn has_block(&self) -> bool {
+        !self.blocks.is_empty()
+    }
+
+    /// Returns the node indexes corresponding to if-else nodes
+    ///
+    pub fn get_ifs(&self) -> &[usize] {
+        &self.ifs
+    }
+
+    /// Returns the `Root` node index of the AST
+    ///
+    pub fn get_root(&self) -> usize {
+        self.root
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Node {
+    /// If-Else node (then branch node indexes, alternative branch node indexes, type of the IfElse node)
+    IfElse(Vec<usize>, Option<Vec<usize>>, TypeOrFuncType),
```

Comment:
It is easier to read these variants with named struct fields (a clue that it might be better as struct fields is if you find yourself writing comments like "(description of field 0, descr of field 1, ...)" to disambiguate the fields):

```suggestion
    /// `if`/`else` node.
    IfElse {
        /// Then branch node indices.
        consequent: Vec<usize>,
        /// Else branch node indices.
        alternative: Option<Vec<usize>>,
        /// The block type for the branches.
        ty: TypeOrFuncType
    },
```